src/Raven.Server/Documents/PeriodicBackup/Retention/RetentionPolicyRunnerBase.cs

```diff
@@ -0,0 +1,145 @@
+
+        public abstract Task<List<string>> GetFiles(string folder);
+
+        public abstract Task DeleteFolder(string folder);
+
+        public async Task Execute()
+        {
+            if (_retentionPolicy == null ||
+                _retentionPolicy.Disabled ||
+                _retentionPolicy.MinimumBackupsToKeep == null && _retentionPolicy.MinimumBackupAgeToKeep == null)
+                return; // no retention policy
+
+            if (_retentionPolicy.MinimumBackupsToKeep <= 0)
+                return;
+
+            try
+            {
+                var folders = await GetFolders();
+
+                var allBackupFolders = new SortedList<DateTime, string>();
+
+                foreach (var folder in folders)
+                {
+                    var folderDetails = RestoreUtils.ParseFolderName(folder);
+                    if (DateTime.TryParseExact(
+                            folderDetails.BackupTimeAsString,
+                            BackupTask.DateTimeFormat,
+                            CultureInfo.InvariantCulture,
+                            DateTimeStyles.None,
+                            out var backupTime) == false)
+                    {
+                        // TODO: log this
+                        // couldn't parse the backup time
+                        continue;
+                    }
+
+                    if (string.Equals(folderDetails.DatabaseName, _databaseName, StringComparison.OrdinalIgnoreCase) == false)
+                    {
+                        // a backup for a different database
+                        continue; // a backup for a different database
+                    }
+
+                    var files = await GetFiles(folder);
+                    var hasFullBackupOrSnapshot = files.Any(IsFullBackupOrSnapshot);
+                    if (hasFullBackupOrSnapshot == false)
+                    {
+                        // no backup files
+                        continue;
+                    }
+
+                    allBackupFolders.Add(backupTime, folder);
+
+                    bool IsFullBackupOrSnapshot(string filePath)
+                    {
+                        var extension = Path.GetExtension(filePath);
+
+                        return Constants.Documents.PeriodicBackup.FullBackupExtension.Equals(extension, StringComparison.OrdinalIgnoreCase) ||
+                               Constants.Documents.PeriodicBackup.EncryptedFullBackupExtension.Equals(extension, StringComparison.OrdinalIgnoreCase) ||
+                               Constants.Documents.PeriodicBackup.SnapshotExtension.Equals(extension, StringComparison.OrdinalIgnoreCase) ||
+                               Constants.Documents.PeriodicBackup.EncryptedSnapshotExtension.Equals(extension, StringComparison.OrdinalIgnoreCase);
+                    }
+                }
+
+
+                var minimumBackupsToKeep = Math.Max(_retentionPolicy.MinimumBackupsToKeep ?? 1, 1);
```

Comment:
Not a big deal, but Math.Max is not needed, because you cannot reach this line when MinimumBackupsToKeep is <= 0

Reply:
done. need to verify that we cannot have MinimumBackupsToKeep <= 0